Tried to clarify the concept of service ports

kolloch opened the pull request.

gkleiman commented on docs/docs/service-discovery-load-balancing.md:
```diff
@@ -6,15 +6,15 @@ title: Service Discovery & Load Balancing
 Once our app is up and running, we need a way to send outside traffic to it. If we're running multiple apps, they also need a way to find each other. We can use [Mesos-DNS](https://github.com/mesosphere/mesos-dns) for service discovery through the domain name system ([DNS](http://en.wikipedia.org/wiki/Domain_Name_System)). Mesos-DNS generates a hostname for each application running on Mesos and t ...
 
-An alternative way for service discovery is to run a TCP proxy on each host in the cluster, and transparently forward a static port on localhost to the hosts that are running the app. That way, clients simply connect to that port, and the implementation details of discovery are completely abstracted away. This approach is sufficient if all apps are launched through Marathon. 
+An alternative way for service discovery is to run a TCP proxy on each host in the cluster, and transparently forward connection to the static service port on localhost to the dynamically assigned host/port combinations of the individual app instances (= *tasks*). That way, clients simply connect to that port, and the implementation details of discovery are completely abstracted away. This approa ...
```
s/connection/connections/

mgummelt replied:
> An alternative way for service discovery is to...

this isn't correct english.  maybe you mean "to implement service discovery"

gkleiman commented on docs/docs/service-discovery-load-balancing.md:
```diff
@@ -6,15 +6,15 @@ title: Service Discovery & Load Balancing
+## Service Ports Assignment
 
-When an application is created in Marathon (either through the REST API or the front end) - you may assign it one or more ports. These ports may be any valid port number or 0, which Marathon will use to randomly assign a port number.
+When an application is created in Marathon (either through the REST API or the front end) - you may assign it one or more ports. These service ports may be any valid port number or 0 which Marathon will use to randomly assign a service port number. Marathon will ensure that all dynamically assigned service ports a unique for this Marathon instance to avoid overlaps. Marathon does not use these as ...
```
s/you may assign it one or more ports/you may assign it one or more service ports/

s/or 0 which Marathon will use to randomly assign a service port number/If 0 is specified, then Marathon will assign a random service port/

s/dynamically assigned service ports a unique for this Marathon/dynamically assigned service ports are unique for this Marathon/

s/to avoid overlaps//

mgummelt replied:
Please default to active voice rather than passive voice.  E.g. "When you create an application" rather than "When an application is created"

> you may assign it one or more ports

This sounds like you're talking about the "ports" property, but really you're only talking about service ports + bridge networking.

Link to the "servicePort" section in the REST API, since that's what you're describing here.  And you should be clear that this only applies to docker + bridge networking.

Does "Marathon instance" mean a "task"?  If so, let's just call them "tasks".  Or does it mean "the marathon framework"?

gkleiman commented on docs/docs/service-discovery-load-balancing.md:
```diff
@@ -6,15 +6,15 @@ title: Service Discovery & Load Balancing
+When an application is created in Marathon (either through the REST API or the front end) - you may assign it one or more ports. These service ports may be any valid port number or 0 which Marathon will use to randomly assign a service port number. Marathon will ensure that all dynamically assigned service ports a unique for this Marathon instance to avoid overlaps. Marathon does not use these as ...
 
-This port is used to ensure no two applications can be run using Marathon with overlapping port assignments (i.e. that the ports are globally unique).
+Note that these service ports should not be confused with the host ports of the individual tasks. Since multiple tasks may run on the same node, each instance (or *task*) is assigned a random host port.
```
The second sentence is not always true.

gkleiman commented on docs/docs/service-discovery-load-balancing.md:
```diff
@@ -6,15 +6,15 @@ title: Service Discovery & Load Balancing
+Note that these service ports should not be confused with the host ports of the individual tasks. Since multiple tasks may run on the same node, each instance (or *task*) is assigned a random host port.
 
-However, this is not the port that each instance of that application is assigned when running on slave nodes within the Mesos cluster. Since multiple instances may run on the same node, each instance (or *task*) is assigned a random port. This can be accessed from the `$PORT` environment variable which is set by Marathon.
+From within the task execution the dynamically assigned host ports can be assessed via environment variables. See [task environment variables - host ports]({{ site.baseurl }}/docs/task-environment-vars.html#host-ports) for more information.
```
s/can be assessed/are available/

kolloch commented:
I hopefully addressed all suggestions. I do know that the documentation (and the API) should be rewritten. Please let's keep this change focussed on making a substantial iterative improvement but not more.

gkleiman commented on docs/docs/service-discovery-load-balancing.md:
```diff
@@ -6,15 +6,37 @@ title: Service Discovery & Load Balancing
+
+## Service Ports Assignment
+
+When Marathon creates an application in Marathon (either through the REST API or the front end),
```
s/When Marathon creates/When you create/

gkleiman commented on docs/docs/service-discovery-load-balancing.md:
```diff
@@ -6,15 +6,37 @@ title: Service Discovery & Load Balancing
+
+When Marathon creates an application in Marathon (either through the REST API or the front end),
+you may assign one or more service ports to it. 
+You can specify an valid port numbers as service ports or you can use 0 to indicate that Marathon should assign
```
s/an valid port numbers/valid port numbers/ ?

Maybe we should specify that a "valid port number" is a number greater than 0 and that they must be unique across all apps.

gkleiman commented on docs/docs/service-discovery-load-balancing.md:
```diff
@@ -6,15 +6,37 @@ title: Service Discovery & Load Balancing
+Note that these service ports should not be confused with the host ports of the individual tasks. 
+Since multiple tasks may run on the same node, host ports are usually assigned dynamically per task if
+you do not use `"requirePorts"` or set the `"hostPort"` to non-zero in the docker `"portMappings"`.
+
```
s/if you do not use `"requirePorts"` or set the `"hostPort"` to non-zero in the docker `"portMappings"/unless you use `"requirePorts"` in the case of a non-Docker app or set the `"hostPort"` to a non-zero value in a Docker app/

kolloch replied:
your suggestion is wrong. portMappings are only used for BRIDGE networking.

gkleiman replied:
What about:

Since multiple tasks may run on the same node, host ports are assigned dynamically per task unless you use `"requirePorts"`, or set the `"hostPort"` to non-zero in the Docker `"portMappings"`. Note: Docker `"portMappings"` are ignored unless the network type is `BRIDGE`.

gkleiman commented on docs/docs/service-discovery-load-balancing.md:
```diff
@@ -6,15 +6,37 @@ title: Service Discovery & Load Balancing
+You can specify an valid port numbers as service ports or you can use 0 to indicate that Marathon should assign
+service ports. 
+
+For applications using docker and `BRIDGE` networking, you assign the servicePorts in the 
```
s/docker/Docker/g

gkleiman commented:
Just a few more comments, thanks for working on this!

kolloch commented:
Updated the pull request again. As a curtesy to the reviewer, here is the diff:

```diff
diff --git a/docs/docs/service-discovery-load-balancing.md b/docs/docs/service-discovery-load-balancing.md
index b30af23..0b1b76c 100644
--- a/docs/docs/service-discovery-load-balancing.md
+++ b/docs/docs/service-discovery-load-balancing.md
@@ -13,12 +13,13 @@ are completely abstracted away. This approach is sufficient if all apps are laun

 ## Service Ports Assignment

-When Marathon creates an application in Marathon (either through the REST API or the front end),
+When you create a new application in Marathon (either through the REST API or the front end),
 you may assign one or more service ports to it.
 You can specify all valid port numbers as service ports or you can use 0 to indicate that Marathon should allocate
-free service ports to the app automatically.
+free service ports to the app automatically. If you do choose your own service port, you have to ensure yourself
+that it is unique across all of your applications.

-For applications using docker and `BRIDGE` networking, you assign the servicePorts in the
+For applications using Docker and `BRIDGE` networking, you assign the servicePorts in the
 `portMappings`. For applications with `HOST` networking or without docker, you use the `ports` attribute
 to assign the service ports. Look at the [REST API docs]({{ site.baseurl }}/docs/rest-api.html#ports-array-of-integers)
 for details.
@@ -32,8 +33,8 @@ It is the responsibility of a separate component to make use of this information
 A sample script that does this is provided with Marathon, the `haproxy-marathon-bridge`.

 Note that service ports should not be confused with the host ports of the individual tasks.
-Since multiple tasks may run on the same node, host ports are usually assigned dynamically per task if
-you do not use `"requirePorts"` or set the `"hostPort"` to non-zero in the docker `"portMappings"`.
+Since multiple tasks may run on the same node, host ports are usually assigned dynamically per task
+unless you use `"requirePorts"` or set the `"hostPort"` to non-zero in the docker `"portMappings"`.
 From within the task execution the dynamically assigned host ports are exposed via environment variables.
 See [task environment variables - host ports]({{ site.baseurl }}/docs/task-environment-vars.html#host-ports) for more information.
```

gkleiman commented:
Thanks a lot for including the diff, it made reviewing much easier.

Added just a last comment, I hope that clarifies how to require/force ports.

kolloch commented:
@gkleiman Can you point me to your comment, please? I don't see it.

kolloch commented on docs/docs/service-discovery-load-balancing.md:
```diff
@@ -6,15 +6,37 @@ title: Service Discovery & Load Balancing
+
+Note that service ports should not be confused with the host ports of the individual tasks. 
+Since multiple tasks may run on the same node, host ports are usually assigned dynamically per task
+unless you use `"requirePorts"` or set the `"hostPort"` to non-zero in the docker `"portMappings"`.
```
I saw your comment in an email but not here. Strange. Anyways: I would rather not mention the Docker BRIDGE thing again here. It is mentioned two paragraphs before.

gkleiman commented:
Merged, bäm, thanks! :+1:
